File: 백준/Platinum/2419. 사수아탕/사수아탕.cc
#include <iostream>
#include <algorithm>
#include <vector>
#include <cstring>
using namespace std;
const int INF = 2147483647;
int n, m;
vector<int> arr;
int dp[301][301];
int ans = 0;

int go(int from, int to, int remain) {
	if (!remain) return 0;
	int&ret = dp[from][to];
	if (ret != -1) return ret;
	
	int l = min(from, to), r = max(from, to);
	ret = INF;
	if (r < n) ret = min(ret, go(l, r + 1, remain - 1) + remain * (arr[r + 1] - arr[to]));
	if (l > 0) ret = min(ret, go(r, l - 1, remain - 1) + remain * (arr[to] - arr[l - 1]));
	return ret;
}

int main() {
	cin.tie(nullptr);
	cout.tie(nullptr);
	ios_base::sync_with_stdio(false);
	cin >> n >> m;
	arr.resize(n + 1);
	for (int i = 1; i <= n; i++)
		cin >> arr[i];
	sort(arr.begin(), arr.end());
	int start = lower_bound(arr.begin(), arr.end(), 0) - arr.begin();
	
	for (int i = 1; i <= n; i++) {
		memset(dp, -1, sizeof(dp));
		ans = max(ans, i * m - go(start, start, i));
	}
	cout << ans << '\n';
}

File: 백준/Platinum/2419. 사수아탕/사수아탕.py
import sys;input = sys.stdin.readline
import math


def lowerbound(li, n):
    start, end = 0, len(li)
    while start < end:
        mid = (start + end) // 2
        if n <= li[mid]:
            end = mid
        else:
            start = mid + 1
    return start


def solve(s, e, candy):
    if candy == 0:
        return 0
    l, r = min(s, e), max(s, e)
    if dp[s][e] != -1:
        return dp[s][e]
    dp[s][e] = math.inf
    if r < n:
        dp[s][e] = min(dp[s][e], solve(l, r + 1, candy - 1) + candy * (li[r + 1] - li[e]))
    if l > 0:
        dp[s][e] = min(dp[s][e], solve(r, l - 1, candy - 1) + candy * (li[e] - li[l - 1]))
    return dp[s][e]


n, m = map(int, input().split())
li = [0] * (n+1)
for i in range(n):
    li[i] = int(input())
li.sort()
s = lowerbound(li, 0)
res = 0
for i in range(1, n+1):
    dp = [[-1] * (n+1) for _ in range(n+1)]
    res = max(res, i * m - solve(s, s, i))

print(res)